src/Storages/MergeTree/MergeTreeDataWriter.cpp

```diff
@@ -245,23 +245,31 @@ Block MergeTreeDataWriter::mergeBlock(const Block & block, SortDescription sort_
 
     IMergingAlgorithm::Input input;
     input.set(std::move(chunk));
-    input.permutation = *permutation;
+    input.permutation = permutation;
 
     IMergingAlgorithm::Inputs inputs;
     inputs.push_back(std::move(input));
     merging_algorithm->initialize(std::move(inputs));
 
     IMergingAlgorithm::Status status = merging_algorithm->merge();
-    while (!status.is_finished)
-        status = merging_algorithm->merge();
+
+    /// Check that after first merge merging_algorithm is waiting for data from input 0.
+    if (status.required_source != 0)
+        return block;
```

Comment:
I think that it would be better to throw exception then hide a possible error.
It is possible to disable using setting anyway.